keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,821 @@
+*   Beta features like[ InPlacePodResizing](https://kubernetes.io/docs/tasks/configure-pod-container/resize-container-resources/) may be enabled in the Control Plane but not on all the Nodes due to version skew. There currently exists a[ workaround](https://github.com/kubernetes/kubernetes/blob/23258f104d74c6f27fd4db94940d745d9d463a8f/pkg/apis/core/validation/validation.go#L5796) to handle this version skew by looking for alternate signals from the pod spec. But having such workarounds may not always be possible. Alternatively, we could use` pod.spec.nodeName` and validate if the Node has the required capability.
+
+
+### Story 2
+
+Prevent Pods from being scheduled on Nodes lacking their required Kubelet, Kernel, or Runtime features, avoiding errors, degraded performance, or silent failures.
+
+**Examples:**
+
+*   Linux kernel can be compiled with support for both AppArmor and SELinux security modules, only one can be actively enforced at any given time. This is a key reason why Kubernetes clusters can exhibit heterogeneity in their security configuration, making it crucial for the `kube-scheduler` to know which Linux Security Module is active on each node (typically via Node Labels) for pod placement. Instead, kubelet can expose this kernel capability via NodeCapabilities and the kube-scheduler can infer this from the pod spec (`securitycontext.seLinuxOptions,securitycontext.appArmorProfile`) and match it to the right Node.
+*  [PodLevelResources](https://kubernetes.io/docs/tasks/configure-pod-container/assign-pod-level-resources/) is currently not supported on Windows Nodes. The capability advertised by kubelet can be dependent on both the feature gate and the operating system which the kube-scheduler can filter out all Windows Nodes.
+*   When pods request for a specific [RunTimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/) (`spec.runtimeClassName`), we currently use node selectors to make sure they are on appropriate nodes . Instead, kube-scheduler can automatically identify nodes that possess the corresponding capability.
+
+## Design Details
+
+### API Changes
+
+ \
+Add a field `NodeCapabilities` field as type `map[string]string `to the` NodeSpec.NodeStatus` structure. 
+
+```
+type NodeStatus struct {
+    // ... existing fields
+    // NodeCapabilities provides a structured way to report various capabilities of the node. Keys are typically DNS-style and values are strings.
+    // +optional 
+    // +featureGate=NodeCapabilities
+    NodeCapabilities map[string]string `json:"capabilities,omitempty"`
+}
+
+
+// Node object remains unchanged in spec, only status is modified.
+type Node struct {
+    ...
+    Status NodeStatus `json:"status,omitempty" `
+}
+
+```
+
+**Example:**
+
+```
+NodeCapabilities:
```

Comment:
So this is pretty much the same thing as labels, except:
- node capabilities are immutable
- they are derived automatically by kubelet instead of being a part of kubelet config

Is my understanding correct?

If so, how can infrastructure autoscaling predict what node capabilities will be available on a node once it is created? Will this require dedicated implementation for each capability?

Reply:
Yes, these are basically just implicit labels & label selectors.

Good question about autoscaling. When we were discussing the criteria for what goes in to capabilities, I proposed the heuristic requirement that all nodes in a NodePool should have the same set of capabilities. We don't actually have an common concept of NodePool in Kubernetes, but I think that should get the intent across. If that is the case, could the autoscaler look at a sample instance in the same pool and assume that scaling nodes with the same template will produce the same set of capabilities?

Reply:
> If that is the case, could the autoscaler look at a sample instance in the same pool and assume that scaling nodes with the same template will produce the same set of capabilities?

That works if such a node exists. In case of NodeAutoProvisioning and creating new node-pool, this won't work...

Reply:
I'm not sure we have any good options here. Would definitely like to hear suggestions. What I can think of:

1. Require an "archetype" instance, basically what I suggested in https://github.com/kubernetes/enhancements/pull/5347#discussion_r2133072894. This prevents scale-to-zero node pools, but as long as an instance is initially created, the NAP can save a snapshot.
2. Define capabilities statically based on the node template. This would require replicating Kubelet admission logic, and knowing the {version, config}->capability mapping for the key components (kernel, runtime).

I don't really think (2) is viable.

Reply:
Thanks all. I captured the above discussion in the "Future Considerations/Cluster Autoscaler Integration" section in the KEP.

cc @SergeyKanzhelev